Adding support for createOrUpdate methods

cmelchior opened the pull request:
This PR extends https://github.com/realm/realm-java/pull/565 with createOrUpdate methods.
Json support has been refactored due to this to be able to support both standalone and managed objects.

The following methods has been added for JSON: 
```
Realm.createOrUpdateObjectFromJson(Class, JSONObject)
Realm.createOrUpdateObjectFromJson(Class, InputStream)
Realm.createOrUpdateObjectFromJson(Class, String)
Realm.createOrUpdateAllFromJson(Class, JSONArray)
Realm.createOrUpdateAllFromJson(Class, InputStream)
Realm.createOrUpdateAllFromJson(Class, String)
```

And for standalone objects
```
Realm.copyToRealmOrUpdate(RealmObject)
Realm.copyToRealmOrUpdate(Iterable RealmObject)
```

Especially the last two gets a bit long winded and could probably use some iteration on naming, other suggestions
```
Realm.createOrUpdate(RealmObject); // From iOS
Realm.addOrUpdate(RealmObject); 
Realm.copyOrUpdateToRealm(RealmObject); // similar naming scheme as JSON methods
Realm.copyOrUpdate(RealmObject); // shorter, would imply that copyToRealm gets renamed to copy()
Realm.updateOrCopyToRealm(RealmObject); // emphasis on updates, which is the primary use case.
Realm.copyToRealm(RealmObject,  Flag); // Flags could be UPDATE, , INSERT
Realm.update(RealmObject); // only updates, return null if object couldn't be updated.
```

astigsen commented:
> Realm.copyToRealmOrUpdate(RealmObject)
> Realm.copyToRealmOrUpdate(Iterable RealmObject)
> Especially the last two gets a bit long winded and could probably use some iteration on naming

How about just ´createOrUpdate()´. No reason to reiterate the that we are adding to a realm, given that the method is called on the realm.

Any specific reason that these methods are not on the RealmObject like in the cocoa binding?

cmelchior commented:
We discarded the use of ```create``` when working with standalone objects, as the only way of getting them into Realm is to create new proxy objects we can return to the user. Copy conveys the intent better that the returned object is different from the standalone object. The downside to that is that ```Realm.copy()``` is really unclear, hence ```Realm.copyToRealm()```.

I am however also a big fan of  ```Realm.create()``` and ```Realm.createOrUpdate()```.

I cannot remember the exact reason for putting them on Realm instead of RealmObject. A couple of guesses:

- The type system is different ie. we cannot prevent this from happening: ```FooClass.copyOrUpdate(BarClassInstance)```. Which would make for an very odd API. Having all those methods on Realm makes it more intuitive that you can throw any RealmObject at them.

- Working with collections get a bit awkward as you *have* to use RealmList. If you keep your standalone objects in any other collection (of which there is many) you have to convert it to a RealmList first.

astigsen commented:
> Copy conveys the intent better that the returned object is different from
the standalone object.

Well, if you call a method that creates a new object from an existing
object, you would probably expect to get the new object back. In some ways
this is even more clear than copy.

*Alexander Stigsen, CEO*
Realm Inc

148 Townsend St
San Fransisco, CA 94107
United States

On Fri, Feb 6, 2015 at 10:49 AM, Christian Melchior <
notifications@github.com> wrote:

> We discarded the use of create when working with standalone objects, as
> the only way of getting them into Realm is to create new proxy objects we
> can return to the user. Copy conveys the intent better that the returned
> object is different from the standalone object. The downside to that is
> that Realm.copy() is really unclear, hence Realm.copyToRealm().
>
> I am however also a big fan of Realm.create() and Realm.createOrUpdate().
>
> I cannot remember the exact reason for putting them on Realm instead of
> RealmObject. A couple of guesses:
>
>    -
>
>    The type system is different ie. we cannot prevent this from
>    happening: FooClass.copyOrUpdate(BarClassInstance). Which would make
>    for an very odd API. Having all those methods on Realm makes it more
>    intuitive that you can throw any RealmObject at them.
>    -
>
>    Working with collections get a bit awkward as you *have* to use
>    RealmList. If you keep your standalone objects in any other collection (of
>    which there is many) you have to convert it to a RealmList first.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/pull/827#issuecomment-73290909>.
>

cmelchior commented:
Some of the reasoning was also that we wanted a different name than Cocoa because the method behaves differently, ie. on Cocoa you convert the existing object, here we return a new object. However that distinction might be lost on people who only uses the API on a single platform.

But it is nice to get other eyes on the naming scheme, as many different viewpoints as possible is good, and I wouldn't be afraid of going back and renaming some of our existing methods if it made those from this PR fit in better.

cmelchior commented:
A few comments:

- Json methods has moved from RealmObject to their own class RealmJsonImpl. This is in order to support both standalone and proxy objects. It also cleans up RealmObject.

- Added Generator class for RealmJsonImpl.

- Refactored all annotation processor utility methods into Utils.

- Current name will be copyToRealmOrUpdate(), we go with readability before brevity.

- Added createOrUpdateJson* for all variants.

- Added unit tests.

Ready for review: @emanuelez @kneth 
Note that the failing unit test is unrelated and will be fixed.

kneth commented on realm-annotations-processor/src/main/java/io/realm/processor/RealmJSonImplGenerator.java:
```diff
@@ -0,0 +1,127 @@
+        writer.emitEmptyLine();
+
+        writer.emitImports(
+                "android.util.JsonReader",
```
Are imports better to be in alphabetic order?

cmelchior replied:
Yup, it has been changed

kneth commented on realm-annotations-processor/src/main/java/io/realm/processor/Utils.java:
```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    public static String joinStringList(List<String> strings, String separator) {
```
Why not use `String.join()`?

kneth commented on realm/src/androidTest/java/io/realm/RealmJsonTest.java:
```diff
@@ -431,4 +433,148 @@ public void testCreateObjectFromJsonStream_nullValues() throws IOException {
+        } catch (IllegalArgumentException expected) {
+            return;
+        }
+        fail();
```
Why not move `fail()` into `try` block?

cmelchior replied:
This way it will fail on anything but the IllegalArgumentException.

kneth commented on realm/src/main/java/io/realm/Realm.java:
```diff
@@ -755,6 +906,33 @@ private static void initializeRealm(Realm realm) {
 
     /**
+     * Tries to update an existing object defined by its primary key with new JSON data. If no existing object could be
+     * found a new object will be saved in the Realm. This must happen inside a transaction.
```
"inside" -> "within"?

According to http://en.wiktionary.org/wiki/within "within" means "before a specific duration ends" and the transaction is the "specific duration".

emanuelez commented on realm/src/main/java/io/realm/Realm.java:
```diff
@@ -1138,6 +1372,7 @@ boolean contains(Class<?> clazz) {
      * @return A sorted RealmResults containing the objects.
      * @throws java.lang.IllegalArgumentException if a field name does not exist.
      */
+    @SuppressWarnings("unchecked")
```
If these methods are tested they should now appear as unchecked. Remove warning suppressor or implement unit test

cmelchior replied:
Even though they are private they are tested intensively through the copyOrUpdate unit tests, so it should be OK to mark them as unchecked.

emanuelez commented on realm/src/main/java/io/realm/Realm.java:
```diff
@@ -1018,6 +1230,28 @@ public void writeEncryptedCopyTo(File destination, byte[] key) throws IOExceptio
+     * @throws java.lang.IllegalArgumentException if RealmObject is {@code null} or doesn't have a Primary key defined.
+     * @see {@link #copyToRealm(Iterable)}
+     */
+    public <E extends RealmObject> List<E> copyToRealmOrUpdate(Iterable<E> objects) {
```
Shouldn't we return a RealmList, so it can be queried if needed?

cmelchior replied:
We should yes, but it is the same as for copyToRealm(Iterable). We need to be able to return a tableview containing all inserted values which currently isn't feasible, at least that was how I remember the conclusion from when we discussed it.
